Multithreading Support on predict_many_sk and predict_many_v2

Snipa22 opened the pull request:
Added multithreading/pool support on the two main generation functions.  This allows for a linear speedup of x/NUM_THREADS as defined in the local_settings.py.

Updated readme.MD to reference the new required setting - May be worth adding as a default to the main settings.py?

Running this in a docker environment with PostgreSQL on a SSD.

predict_many_v2 is still running in the background right now, once it's done, I'll provide some time stats.

Machine Stats:
2x Intel(R) Xeon(R) CPU           L5639  @ 2.13GHz
72Gb ram

Runtimes:
Running with 24 threads:
- predict_many_sk:
  - real 2m 7.850s
  - user 27m 25.829s
  - sys 1m 41.492s

Non-threaded:
- predict_many_sk:
  - real 32m 2.854s
  - user 22m 5.785s
  - sys 0m 44.857s

Snipa22 committed:
Update to setup multiprocesssing pools for better speed.

Snipa22 committed:
It's NUM_THREADS not MAX_THREADS

Snipa22 committed:
Updates to move the functions somewhere usable...

Snipa22 committed:
Cleanup/Grammar and per-line printing

owocki commented on history/management/commands/predict_many_v2.py:
```diff
@@ -1,6 +1,30 @@
+from multiprocessing import Pool
+
+
+def do_classifier_test(ticker, hidden_layers, min_back, epochs, granularity, datasetinputs,
```
s/classifier_test/prediction_test/g

Snipa22 committed:
Corrected the prediction tests to the correct function names

Snipa22 committed:
Lets try that again.  Corrected the function names.

owocki commented:
merging tomorrow. speak now if comments, problems... etc etc

owocki commented:
/Tip 0.1 BTC

igorpejic commented on readme.md:
```diff
@@ -153,6 +153,9 @@ SMTP_USERNAME = '<smtp_user>'
 LOG_FILE = '/var/log/django.log'
+
+# Configuration of the number of threads to be used for predictions - Set to CPU cores + 1 if dedicated machine
+NUM_THREADS = 5
```
Can we add this to the main `settings.py` with a default value like 2 in order to avoid breaking functionality for those that won't update their `local_settings`.
Another option is to cleverly catch the broken import of the settings and notify the user he needs to update the `local_settings`, but I'm kinda leaning on the first option.

owocki replied:
looks like this is done.

t0mk commented:
Maybe NUM_THREADS could default to multiprocessing.cpu_count() or sth like this https://github.com/thefab/cpu_cores/blob/master/cpu_cores/linux.py

t0mk commented:
I rebased this on my Docker branhc and it works. I can see the two python threads spinning the two CPU on the crappy DO server.

Snipa22 commented:
I'd considered moving this to an automated number of cpus, but considering that this thing can and will thrash your entire system if you set it too high, and/or have other things running, my system was choking a bit at 24, because the DB and everything else was shared on those 24 cores.

Snipa22 commented:
Set the default worker thread count to 1, this will keep things from breaking for people who don't read the readme!

Snipa22 committed:
Setting default number of worker threads to 1.
Will maintain current functionality/speed, unless overriden in local_settings.py

owocki commented:
thanks, merging

darkdarkdragon commented:
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__This issue now has a funding of 0.1 ETH (69.21 USD @ $692.1/ETH) attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](http://localhost:8000/issue/owocki/pytrader/42/475).
 * Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $0 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

darkdarkdragon commented:
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 0.2 ETH (138.42 USD @ $692.1/ETH) attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](http://localhost:8000/issue/owocki/pytrader/42/479).
 * Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $0 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)
